src/Symfony/Component/Validator/Constraints/ExpressionValidator.php

```diff
@@ -44,7 +45,12 @@ public function validate($value, Constraint $constraint)
 
         $variables = array();
         $variables['value'] = $value;
-        $variables['this'] = $this->context->getObject();
+
+        if ($constraint->dataPath) {
+            $variables['this'] = PropertyAccess::createPropertyAccessor()->getValue($this->context, $constraint->dataPath);
```

Comment:
Still I think we can get rid of `dataPath` altogether and set the root context directly. Even, if the root is a form instance then we could set the form data.

Reply:
How would you do that if the data is not available at the time you configure the constraint (which by the way is almost always the case when not using PHP to set up your constraints)?

Reply:
I'm not sure to understand your question :( but in short this is what I meant:
```php
if (null === $variables['this'] = $this->context->getObject()) {
    if (($root = $this->context->getRoot()) instanceof FormInterface) {
        $variables['this'] = $root->getData();
    } else {
        $variables['this'] = $root;
    }
}
```

Reply:
But this would force the Validator component to know something about the Form component. I don't think that's a good idea. There may also be other similar issues (not related to Symfony) which we couldn't fix here.

Reply:
Right, then just `$variables['this'] = $this->context->getRoot()`?

Reply:
IMHO there is not need to configure a `dataPath` if we already have a data root, why not start from it directly (`this["foo"].getData()`)?